pkg/workflow/controller/handlers/workflowrun/handler.go

```diff
@@ -84,11 +104,45 @@ func (h *Handler) ObjectUpdated(obj interface{}) {
 	// the GC queue.
 	h.GCProcessor.Add(originWfr)
 
-	// If the WorkflowRun has already been terminated(Completed, Error, Cancel) or waiting for external events, skip it.
+	// If the WorkflowRun has already been terminated(Completed, Error, Cancel), send notifications if necessary,
+	// otherwise directly skip it.
 	if originWfr.Status.Overall.Status == v1alpha1.StatusCompleted ||
 		originWfr.Status.Overall.Status == v1alpha1.StatusError ||
-		originWfr.Status.Overall.Status == v1alpha1.StatusWaiting ||
 		originWfr.Status.Overall.Status == v1alpha1.StatusCancelled {
+		// No need to send notifications for workflowruns finished before workflow controller starts.
+		if originWfr.Status.Overall.LastTransitionTime.Before(controllerStartTime) {
+			return
+		}
+
+		// Send notifications with workflowrun.
```

Comment:
Could you please extract this *sending notification code block* to a method ?

Reply:
Done